Register services in service registry during the service bootstrap phase

Kami opened the pull request:
This pull request updates service bootstrap phase (``st2common.service_setup.common_setup`` function) to register every StackStorm service in the service registry.

### Background and Context

Registering each service in a service segistry will allow us to have a better overview of the active and running services. This will allow us to implement things such as more dynamic (utilization and capability based) action execution scheduling / routing and more.

The goal was so it also works outside of the Kubernetes context in a traditional HA environment. It's also worth noting that Kubernetes has it's own service registry concept, but that's based on a slightly higher level Kubernetes view of a service (two related concepts, but not exactly the same).

### Implementation Details

Implementation utilizes [group membership](https://docs.openstack.org/tooz/queens/user/tutorial/group_membership.html) primitives from the [OpenStack tooz library](https://docs.openstack.org/tooz/latest/).

There were multiple reasons (as discussed before) for utilizing tooz library:

1. Consistency

We already use it in other places for distributed locking.

2. Support for multiple drivers / adapters

We will officially probably only support a single backend (e.g. ZooKeeper or etcd), but people will still be able to user other tooz supported backends at their own risk and this will also help with testing.

Keep in mind that only the following tooz drivers support group membership primitives which we utilize: zookeeper, redis, etcd3, etcd3gw, memcached.

To be able to implement it, I needed to change ``coordinator_setup`` and ``get_coordinator`` function signature so we can pass ``start_heartbeat`` argument to it. This way coordination service automatically starts periodic heartbeats which tell the backend that a particular service (member) is still alive.

In addition to that, I added ``tools/list_group_members.py`` script which allows user to list active services.

### Terminology

OpenStack tooz library exposes a couple of different terms. Here I describe how they fit into our implementation.

1. Group

Each service is part of a single group. In our case, group represents a service type (e.g. ``api``, ``actionrunner``, ``scheduler``, etc.). To view all the running services of a particular type, we just need to list all the members in a particular group.

2. Member

A service is represented as a member which is part of a particular group.

3. Capabilities

Optional metadata associated with a service (service name, listen host / port, service type, hostname, pid and other info). In the future, action runners will also utilize this to advertise their capabilities (e.g. linux, windows, etc.).

### Example Output

Example output of ``tools/list_group_members.py`` script when all the services are running using launchdev.sh script (to test it locally, I used the redis backend).

```bash
$ python ./tools/list_group_members.py --config-file conf/st2.dev.conf 
Available groups (13):
 - actionrunner
 - garbagecollector
 - workflow_engine
 - resultstracker
 - stream
 - timer_engine
 - auth
 - sensorcontainer
 - api
 - scheduler
 - rulesengine
 - notifier

Members in group "actionrunner" (1):
 - localhost_21837 (capabilities={'type': 'passive', 'hostname': 'localhost', 'pid': 21837, 'name': 'actionrunner'})
Members in group "garbagecollector" (1):
 - localhost_21841 (capabilities={'type': 'passive', 'hostname': 'localhost', 'pid': 21841, 'name': 'garbagecollector'})
Members in group "workflow_engine" (1):
 - localhost_21834 (capabilities={'type': 'passive', 'hostname': 'localhost', 'pid': 21834, 'name': 'workflowengine'})
Members in group "resultstracker" (1):
 - localhost_21871 (capabilities={'type': 'passive', 'hostname': 'localhost', 'pid': 21871, 'name': 'resultstracker'})
Members in group "stream" (1):
 - localhost_21930 (capabilities={'listen_host': '127.0.0.1', 'name': 'stream', 'listen_port': 9102, 'hostname': 'localhost', 'pid': 21930, 'type': 'active'})
Members in group "timer_engine" (1):
 - localhost_21853 (capabilities={'type': 'passive', 'hostname': 'localhost', 'pid': 21853, 'name': 'timerengine'})
Members in group "auth" (1):
 - localhost_21927 (capabilities={'listen_host': '127.0.0.1', 'name': 'auth', 'listen_port': 9100, 'listen_ssl': False, 'hostname': 'localhost', 'pid': 21927, 'type': 'active'})
Members in group "sensorcontainer" (1):
 - localhost_21849 (capabilities={'type': 'passive', 'hostname': 'localhost', 'pid': 21849, 'name': 'sensorcontainer'})
Members in group "api" (1):
 - localhost_21925 (capabilities={'listen_host': '127.0.0.1', 'name': 'api', 'listen_port': 9101, 'hostname': 'localhost', 'pid': 21925, 'type': 'active'})
Members in group "scheduler" (1):
 - localhost_21846 (capabilities={'type': 'passive', 'hostname': 'localhost', 'pid': 21846, 'name': 'scheduler'})
Members in group "rulesengine" (1):
 - localhost_21850 (capabilities={'type': 'passive', 'hostname': 'localhost', 'pid': 21850, 'name': 'rulesengine'})
Members in group "notifier" (1):
 - localhost_21870 (capabilities={'type': 'passive', 'hostname': 'localhost', 'pid': 21870, 'name': 'notifier'})
```

## TODO

- [x] Changelog entry
- [x] Tests

Resolves #4015.

m4dcoder commented on tools/list_group_members.py:
```diff
@@ -0,0 +1,72 @@
+    do_register_cli_opts(cli_opts)
+    config.parse_args()
+
+    main(group_id=cfg.CONF.group_id)
```
Should we add an admin only API endpoint and a command in the st2 CLI to list services, members and capabilities?

Kami replied:
That's a good idea - I was thinking about it.

I will add it and also add a corresponding CLI command.

m4dcoder commented on st2common/st2common/service_setup.py:
```diff
@@ -165,9 +173,39 @@ def setup(service, config, setup_db=True, register_mq_exchanges=True,
     metrics_initialize()
 
+    # Register service in the service registry
+    if service_registry:
```
Should we add a config check here?  If user specifies a backend with no service registry support, we log and warn users and start st2 w/o service registry?

m4dcoder replied:
Did you address this?  Or is this a bad idea?

Kami replied:
We already log a warning when ``get_coordinator`` is called inside ``register_service_in_service_registry`` function.

nmaludy commented:
Should the heartbeat be started for the coordinator/locking driver instances? I believe in some implementations that if locks don't get a heartbeat then they are released.

Kami commented:
@nmlaudy I think you are correct, good catch :+1: 

I will also make that change, it will likely require test changes though. I already have some issues with this change (tests are getting stuck, even though I correctly shut down the coordinator everywhere, could also be a bug in a test driver we use).

Kami commented:
@m4dcoder This PR is now ready for a review.

I've added and fixed tests (there was an issue with tests hanging / timing out) and new CLI commands ``st2 service-registry group list``, ``st2 service-registry member list [--group-id=<group id>]``).

Adding CLI commands was a bit PITA since those commands don't follow common "pack resource" pattern and in general that CLI and http client related code is not the prettiest of things.

Kami commented:
On a related note^ - I propose changing the following RBAC CLI commands in a separate PR in the future:

* ``st2 role list`` -> ``st2 rbac role list``
* ```st2 role-assignment list`` -> ``st2 rbac role-assignment list``

This way it will be immediately obvious those commands refer to RBAC. In the past, we didn't have a pattern established yet for double nested commands so we went with a simple top-level approach which is somewhat confusing.

Kami commented:
Just a heads up - I also pushed a change so we start heartbeat process also in places where we use coordination backend only for locking - 98b77d1.

This is a safer default since it means it will work correctly with backends which are timeout based.

nmaludy commented:
Closes #4015

m4dcoder commented on st2api/st2api/controllers/v1/service_registry.py:
```diff
@@ -0,0 +1,77 @@
+        coordinator = coordination.get_coordinator()
+
+        group_ids = list(coordinator.get_groups().get())
+        group_ids = [group_id_.decode('utf-8') for group_id_ in group_ids]
```
Is the underscore at the end of `group_id_` mean something?

Kami replied:
Yes, ``group_id`` variable is already defined in the method signature and I don't want to re-define / overwrite it in a list comprehension.

m4dcoder replied:
Practice wise, we don't usually put an underscore at the end of a variable.  This is confusing and error prone. Can you rename `group_id_` to something else?  Maybe `item` just for this list comprehension.

m4dcoder commented on st2client/st2client/models/service_registry.py:
```diff
@@ -0,0 +1,48 @@
+
+class ServiceRegistry(core.Resource):
+    _alias = 'service-registry'
+    _display_name = 'service registry'
```
Please capitalize first letters.

Kami replied:
This attribute is really just a place holder, it's not used anywhere.

The class (model) is just needed to get ``st2 service-registry foo bar`` command structure to work since our, not so ideal CLI code, depends on it.

I can remove unused attributes all together.

m4dcoder replied:
Ok these attributes are still here.  So, placeholder or not, please capitalize `service registry` to `Service Registry`. This is so the values are consistent with other models.

m4dcoder commented on st2common/st2common/services/coordination.py:
```diff
@@ -51,16 +55,30 @@ def heartbeat(self):
         return True
 
 
+class NoOpAsyncResult(object):
```
Can you provide a short description why we need to wrap result with this class?

Kami replied:
It's to correctly mimic tooz API - most of the tooz methods return an async result (future).

I can add that as a docstring.

m4dcoder commented on st2common/st2common/services/coordination.py:
```diff
@@ -161,22 +192,44 @@ def coordinator_setup():
 
+    if not use_cache:
+        coordinator = coordinator_setup(start_heart=start_heart)
+        return coordinator
```
return in one line?  `return coordinator_setup(start_heart=start_heart)`

Kami replied:
That's a personal preference - I can prefer return in a separate line.

m4dcoder replied:
Ok. My suggestion here is because we are not doing anything with the variable `coordinator` so why not just return here.  This is also 1 line of code less.  What's the reasoning for your preference?

m4dcoder commented on st2common/st2common/service_setup.py:
```diff
@@ -170,9 +180,57 @@ def setup(service, config, setup_db=True, register_mq_exchanges=True,
+    # Include common capabilities such as hostname and process ID
+    proc_info = system_info.get_process_info()
+    capabilities['hostname'] = proc_info['hostname']
+    capabilities['pid'] = proc_info['pid']
```
Doesn't get_member_id already included this info?  Maybe doesn't matter much, but seems like we're calling system_info.get_process_info twice.

Kami replied:
Nope, this function doesn't include member info. Just hostname and pid.

m4dcoder commented on tools/launchdev.sh:
```diff
@@ -68,6 +68,7 @@ function init(){
     fi
 
     VIRTUALENV=${VIRTUALENV_DIR:-${ST2_REPO}/virtualenv}
+    VIRTUALENV=$(readlink -f ${VIRTUALENV})
```
I assume this is to accommodate symlink in your dev env.  Will this break existing setup for us?

Kami replied:
It's so an absolute path is used everywhere (just resolves relative path to an absolute one). Everything should still work the same.

m4dcoder commented on tools/launchdev.sh:
```diff
@@ -77,6 +78,8 @@ function init(){
         ST2_CONF=${ST2_REPO}/conf/st2.dev.conf
     fi
+
+    ST2_CONF=$(readlink -f ${ST2_CONF})
```
I assume this is to accommodate symlink in your dev env. Will this break existing setup for us?

Kami commented:
@m4dcoder 

> Practice wise, we don't usually put an underscore at the end of a variable. This is confusing and error prone. Can you rename group_id_ to something else? Maybe item just for this list comprehension.

That's a relatively common pattern in Python world. Also when variable names would clash with built-ins such as ``id``, etc.
